src/main/scala/com/microsoft/partnercatalyst/fortis/spark/sinks/cassandra/aggregators/HeatmapOfflineAggregator.scala

```diff
@@ -0,0 +1,99 @@
+      val mentionCount = a.mentioncount + b.mentioncount
+      val sentimentNumerator = a.avgsentimentnumerator + b.avgsentimentnumerator
+      a.copy(
+        mentioncount = mentionCount,
+        avgsentimentnumerator = sentimentNumerator
+      )
+    }).values
+  }
+
+  private def aggregateTileBuckets(tiles: RDD[HeatmapTile], keyspace: String): DataFrame = {
+    import session.implicits._
+    val tilesComputed = tiles.keyBy(r=>{(
+      r.period,
+      r.periodtype, r.perioddate,
+      r.conjunctiontopic1, r.conjunctiontopic2, r.conjunctiontopic3,
+      r.tileid, r.tilez,
+      r.pipelinekey, r.externalsourceid
+    )}).reduceByKey((a,b)=>{
+      val mentionCount = a.mentioncount + b.mentioncount
+      val sentimentNumerator = a.avgsentimentnumerator + b.avgsentimentnumerator
+      a.copy(
+        mentioncount = mentionCount,
+        avgsentimentnumerator = sentimentNumerator
+      )
+    }).values.map(CassandraTileBucket(_))
+
+    val tileBucketDF = tilesComputed.toDF()
+
+    tileBucketDF.createOrReplaceTempView("aggregatedtiles")
+
+    val computedTilesSourceDF = session.read.format("org.apache.spark.sql.cassandra")
+      .options(Map("keyspace" -> keyspace, "table" -> ComputedTilesTable))
+      .load()
+
+    computedTilesSourceDF.createOrReplaceTempView(ComputedTilesTable)
+
+    val cassandraSave = session.sql(IncrementalUpdateQuery)
+
+    cassandraSave
+  }
+
+  private def IncrementalUpdateQuery: String = {
+    val GroupedBaseColumnNames = Seq("periodtype", "period", "conjunctiontopic1", "conjunctiontopic2", "conjunctiontopic3", "perioddate", "tileid", "tilez", "pipelinekey", "externalsourceid").mkString(",a.")
+
+    s"SELECT a.$GroupedBaseColumnNames, " +
+    s"       a.mentioncount + IF(IsNull(b.mentioncount), 0, b.mentioncount) as mentioncount, " +
+    s"       a.avgsentimentnumerator + IF(IsNull(b.avgsentimentnumerator), 0, b.avgsentimentnumerator) as avgsentimentnumerator " +
+    s"FROM   aggregatedtiles a " +
+    s"LEFT OUTER JOIN ${ComputedTilesTable} b " +
+    s" ON a.pipelinekey = b.pipelinekey and a.periodtype = b.periodtype and a.period = b.period " +
+    s"    and a.externalsourceid = b.externalsourceid and a.conjunctiontopic1 = b.conjunctiontopic1 " +
+    s"    and a.conjunctiontopic2 = b.conjunctiontopic2 and a.conjunctiontopic3 = b.conjunctiontopic3 " +
+    s"    and a.tileid = b.tileid and a.tilez = b.tilez"
+  }
+
+  override def aggregateAndSave(events: RDD[Event], keyspace: String): Unit = {
+    val tiles = aggregate(events).cache()
+    tiles.count() match {
+      case 0 => return
+      case _ => {
+        implicit val rowWriter = SqlRowWriter.Factory
+        tiles.saveToCassandra(keyspace, "heatmap")
+
+        val tileBuckets = aggregateTileBuckets(tiles, keyspace).cache()
+        tileBuckets.write
+          .format("org.apache.spark.sql.cassandra")
+          .mode(SaveMode.Append)
+          .options(Map("keyspace" -> keyspace, "table" -> ComputedTilesTable)).save
```

Comment:
We've seen this call be up to 3x slower than using saveToCassandra on the RDD. I wonder if we can use the Cassandra connector's native join capability (via [leftJoinWithCassandra](https://github.com/datastax/spark-cassandra-connector/blob/72c4bc6703b71854746d35a8f8cc2108f48f6730/spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/RDDFunctions.scala#L175-L198) or [joinWithCassandra](https://github.com/datastax/spark-cassandra-connector/blob/72c4bc6703b71854746d35a8f8cc2108f48f6730/spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/streaming/DStreamFunctions.scala#L112-L140)) on our RDD/DStream to get rid of this save call and stick entirely in the saveToCassandra world (this would also help with consistency with the rest of the sink code).

Reply:
100% agree and I had the intention of converting this over to a RDD join. Considering you're OOO, I was going to ask @jcjimenez to run with converting the left outer join from dataframes to RDD.

Reply:
Just FYI: I'm back in action this week and next :)